test/e2e/storage/drivers/base.go

```diff
@@ -0,0 +1,143 @@
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/storage/testpatterns"
+)
+
+// TestDriver represents an interface for a driver to be tested in TestSuite
+type TestDriver interface {
+	// GetDriverInfo returns DriverInfo for the TestDriver
+	GetDriverInfo() *DriverInfo
+	// CreateDriver creates all driver resources that is required for TestDriver method
+	// except CreateVolume
+	CreateDriver(testpatterns.TestVolType)
+	// CreateDriver cleanup all the resources that is created in CreateDriver
+	CleanupDriver(testpatterns.TestVolType)
+	// SkipUnsupportedTest skips test in Testpattern is not suitable to test with the TestDriver
+	SkipUnsupportedTest(testpatterns.TestPattern)
+}
+
+// PreprovisionedVolumeTestDriver represents an interface for a TestDriver that has pre-provisioned volume
+type PreprovisionedVolumeTestDriver interface {
+	TestDriver
+	// CreateVolume creates a pre-provisioned volume.
+	CreateVolume(testpatterns.TestVolType)
+	// DeleteVolume deletes a volume that is created in CreateVolume
+	DeleteVolume(testpatterns.TestVolType)
+}
+
+// InlineVolumeTestDriver represents an interface for a TestDriver that supports InlineVolume
+type InlineVolumeTestDriver interface {
+	PreprovisionedVolumeTestDriver
+	// GetVolumeSource returns a volumeSource for inline volume.
+	// It will set readOnly and fsType to the volumeSource, if TestDriver supports both of them.
+	// It will return nil, if the TestDriver doesn't support either of the parameters.
+	GetVolumeSource(readOnly bool, fsType string) *v1.VolumeSource
+}
+
+// PreprovisionedPVTestDriver represents an interface for a TestDriver that supports PreprovisionedPV
+type PreprovisionedPVTestDriver interface {
+	PreprovisionedVolumeTestDriver
+	// GetPersistentVolumeSource returns a PersistentVolumeSource for pre-provisioned Persistent Volume.
+	// It will set readOnly and fsType to the PersistentVolumeSource, if TestDriver supports both of them.
+	// It will return nil, if the TestDriver doesn't support either of the parameters.
+	GetPersistentVolumeSource(readOnly bool, fsType string) *v1.PersistentVolumeSource
+}
+
+// DynamicPVTestDriver represents an interface for a TestDriver that supports DynamicPV
+type DynamicPVTestDriver interface {
+	TestDriver
+	// GetDynamicProvisionStorageClass returns a StorageClass dynamic provision Persistent Volume.
+	// It will set fsType to the StorageClass, if TestDriver supports it.
+	// It will return nil, if the TestDriver doesn't support it.
+	GetDynamicProvisionStorageClass(fsType string) *storagev1.StorageClass
+}
+
+// DriverInfo represents a combination of parameters to be used in implementation of TestDriver
+type DriverInfo struct {
+	Name       string // Name of the driver
+	FeatureTag string // FeatureTag for the driver
+
+	MaxFileSize        int64           // Max file size to be tested for this driver
+	SupportedFsType    map[string]bool // Map of string for supported fs type
```

Comment:
nit: Instead of `map[string]bool` can we use `sets.String` instead? Its the apimachinery type that pretends to be a `set` since go doesn't have them.

Reply:
@davidz627 

Thank you for your pointing it out. I will fix it as suggested.